### PR TITLE
partially ported code to pystackexchangeAPI usage

### LINKED ISSUE #178: Use pystackexchange API to fetch question and answers
Use https://github.com/lucjon/Py-StackExchange to fetch the question and answer details instead of crawling StackOverflow. It will reduce future errors.

Example: https://github.com/lucjon/Py-StackExchange/blob/master/demo/question.py

@tranchikhang if you are free please have a look into this.

### PR DESCRIPTION
**Pull request by therajdeepbiswas**
resolve #178

we have the following changes (due to the API) -
1. now, question was asked "on January 12, 2017" instead of "3 years, 5 months ago". but if the latter is still preferred, let me know, i can take a shot at a code that calculates the duration passed.
2. same about the last active date
3. we lost rounded off functionality; 126585 views would show up as "126585" instead of "127k". (can attempt to get back previous functionality, as well, just let me know).

note: i think this has caused stronger overheads and the loading time had increased significantly on my local. can anyone else please confirm this?

I'm yet to modify the get_questions_for_query function, once I have approval on the google search function.

@gautamkrishnar @hedythedev, please review

### CODE REVIEW

**Comment by gautamkrishnar** [comment]
I will check the performance implications and let you know.

**Comment by gautamkrishnar** [comment]
I think the problem lies in google search, instead of giving a link to the question sometimes google directly links to the answer. For example https://www.google.com/search?q=site:www.stackoverflow.com+for+loop+python

Following are the links available:
```
https://www.stackoverflow.com/a/47720685
http://www.stackoverflow.com/Questions/89178/in-python-what-is-the-fastest-algorithm-for-removing-duplicates-from-a-list-so
https://www.stackoverflow.com/a/48305616
http://www.stackoverflow.com/a/44063325
http://www.stackoverflow.com/Questions/4039879/best-way-to-find-the-months-between-two-dates
...etc
```
@therajdeepbiswas We should think about a better implementation, we should avoid crawling the StackOverflow that is the only way we can improve performance. Please let me know if you have any ideas. I am currently reading through the StackExchange API to find a solution. 
CC: @hedythedev

File: socli/search.py [hunk 1]
```diff
@@ -24,6 +29,15 @@
 google_search_url = "https://www.google.com/search?q=site:www.stackoverflow.com+"  # Google search query URL
 
 
+def get_question_by_url(question_url):
+    upper = 9 # Currently anticipated maximum question_ID length. Example: 9 digits being numbers upto hundred millionth places
+    question_page_url = requests.get(question_url).url
```
Comment (gautamkrishnar) [inline socli/search.py:34]: We should avoid this `requests.get`

**Comment by gautamkrishnar** [comment]
There is a https://api.stackexchange.com/docs/answers-by-ids API available. It can be used as the following:
For the first URL: https://www.stackoverflow.com/a/47720685
```
import stackexchange
site = stackexchange.Site(stackexchange.StackOverflow)
site.be_inclusive()
ans = site.answer('46767475')
```

I tried the above but it was not working.

**Comment by hedythedev** [comment]
> I tried the above but it was not working.
> 
> Opened a question here: https://meta.stackexchange.com/questions/349379/why-do-stackoverflow-com-a-answer-id-generated-by-google-not-working-when-pass

Someone answered your question saying the answer was deleted, I don't have my computer right now, but I can try another question and answer as soon as I have time.

**Comment by gautamkrishnar** [comment]
@hedythedev yea saw that. So as of now there's no way to get those answers other than crawling the stack overflow site.  It's really unfortunate.

**Comment by therajdeepbiswas** [comment]
@gautamkrishnar @hedythedev
I looked around a little more and turns out that if you have an `answer_id` you can actually pass it into the `answer()` method and use the `answer` object's `question_id` as follows `site.answer(answer_id).question_id` (instead of passing an answer ID into a `question()`, which by design, won't work). That newly obtained `question_id`, again, can be passed into the `question()` method, which indeed returns the correct question. This can be a workaround for when we are receiving inconsistent forms of URLs; (some a `question/` and others `a/`). However, this doesn't yet solve the problem of deleted answers, I'm still looking into it.

**Comment by therajdeepbiswas** [comment]
@gautamkrishnar I followed the same [google search URL](https://www.google.com/search?q=site:www.stackoverflow.com+for+loop+python) as you had posted in your question description and turns out there are different results; ones that completely omit the previous deleted answer URL, for example `a/47720685` has completely disappeared now.

image

I think this is a 'fault' (if I may) on google's side that they're returning URLs that lead deleted answers which then stackoverflow has to handle by redirecting correctly (depending upon whether the user is 10k+ rep or not). I propose one of two alternatives -

1. We remove the fetching of deleted answer URLs completely; we do not try to get the redirect URL at all
2. We handle all URLs differently; if it's a `question/` or a `a/` URL to a non-deleted answer, we directly use the API. If not; **only then** we try to fetch the redirect URL, by the means of `requests.get()`. That I believe will significantly reduce the overhead with the belief that deleted answer results will only show up 5-10% of times that we search.

Let me know if any of these two ways sound good to you, I will start implementing once I have an approval
CC: @hedythedev

**Comment by gautamkrishnar** [comment]
@therajdeepbiswas passing the question id received from the answer API call is okay and more performance efficient compared to parsing and fetching a large HTML IMO. Not sure it is still 2 network calls, maybe it will be slower. As of now, if there's no solution we can abandon this change. In the future, if something changes, we can re-implement this.

**Comment by gautamkrishnar** [comment]
> @gautamkrishnar I followed the same [google search URL](https://www.google.com/search?q=site:www.stackoverflow.com+for+loop+python) as you had posted in your question description and turns out there are different results; ones that completely omit the previous deleted answer URL, for example `a/47720685` has completely disappeared now.
> 
> image
> 
> I think this is a 'fault' (if I may) on google's side that they're returning URLs that lead deleted answers which then stackoverflow has to handle by redirecting correctly (depending upon whether the user is 10k+ rep or not). I propose one of two alternatives -
> 
> 1. We remove the fetching of deleted answer URLs completely; we do not try to get the redirect URL at all
> 2. We handle all URLs differently; if it's a `question/` or a `a/` URL to a non-deleted answer, we directly use the API. If not; **only then** we try to fetch the redirect URL, by the means of `requests.get()`. That I believe will significantly reduce the overhead with the belief that deleted answer results will only show up 5-10% of times that we search.
> 
> Let me know if any of these two ways sound good to you, I will start implementing once I have an approval
> CC: @hedythedev

1. **We remove the fetching of deleted answer URLs completely; we do not try to get the redirect URL at all:** this will break the functionality of Google search. Even though Google is returning the deleted answer URLs the SO site is still pointing the user to the correct question via redirection. Google usually displays the most relevant StackOverflow page first so this approach is not good.
2.  **We handle all URLs differently; if it's a question/ or a a/ URL to a non-deleted answer, we directly use the API. If not; only then we try to fetch the redirect URL, by the means of requests.get(). That I believe will significantly reduce the overhead with the belief that deleted answer results will only show up 5-10% of times that we search.:** This solution looks good 👍 

#### This will be the implementation
1> Search on google
2.1> If the URL has `question/id` use this to fetch question using api.question(id)
2.2>  If the URL has `a/id` use this to fetch question using api.answer(id)
2.2.1> If the api.answer(id) fails use crawling to fetch the answer

**Comment by gautamkrishnar** [comment]
@therajdeepbiswas can you try comparing the performance of passing the correct question id received by calling the answers API with the current version of SOCLI. You can use time command line if you are using Linux.

```bash
time python -m ./socli <some query that returns correct answer id as first result>
```

**Comment by hedythedev** [comment]
Using `time` can lead to different results even using same queries. I think if we really want to test the performance, we should use a timing library (like `timeit` or some other profilers) to produce the best results to compare the methods.

**Comment by therajdeepbiswas** [comment]
@gautamkrishnar @hedythedev take a look at this. I do not understand why entire different questions are being fetched.

**Comment by therajdeepbiswas** [comment]
> 1. We remove the fetching of deleted answer URLs completely; we do not try to get the redirect URL at all

PS: Just for the sake of experiment, I implemented (1) and it is still causing massive overheads. Turns out its not just the `requests.get()` that is causing the overheads, the API calls are lesser so but similarly expensive nonetheless.

@gautamkrishnar we might have to drop this idea. Let me know if you want *just* the `get_stats()` implementation changed, **although**, even that will require to fetch the question ID and then `question` object states like `.score`, etc and ultimately cause overheads again.

**Comment by hedythedev** [comment]
😕 sorry I accidentally clicked approve changes but what I meant was the `import stackexechange` is fine now

**Comment by gautamkrishnar** [comment]
> @gautamkrishnar @hedythedev take a look at this. I do not understand why entire different questions are being fetched.

If you meant about google search that's something that we don't have control. Only thing we can do is to assume that Google is returning the right result that's relevant to the user.

**Comment by gautamkrishnar** [comment]
@hedythedev @therajdeepbiswas Thanks a lot for your time, it's better to drop this idea for now because of the issue with API.

**Pull request closed without merging**